File: c++/82. Remove Duplicates from Sorted List II.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode* deleteDuplicates(ListNode* head) {
        ListNode* new_head = new ListNode(0);
        ListNode* p = head, *pre = new_head; pre->next = p;
        while (p)
        {
            while (p->next && pre->next->val == p->next->val)
            {
                p = p->next;
            }
            
            if (pre->next == p)
                pre = pre->next;
            else
                pre->next = p->next;
            p = p->next;
        }
        
        return new_head->next;
    }
};

File: python/82. Remove Duplicates from Sorted List II.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    def deleteDuplicates(self, head):
        """
        :type head: ListNode
        :rtype: ListNode
        """
        dummy = ListNode(0)
        p = dummy

        while head:
            tmp = head.next
            f = False
            while tmp and tmp.val==head.val:
                tmp = tmp.next
                f = True
            if not f:
                p.next = head
                p = p.next
                p.next = None                
            head = tmp
        return dummy.next